lib/ecto/enum.ex

```diff
@@ -213,35 +213,47 @@ defmodule Ecto.Enum do
     "#Ecto.Enum<values: #{inspect Keyword.keys(mappings)}>"
   end
 
-  @doc "Returns the possible values for a given schema and field"
-  @spec values(module, atom) :: [atom()]
-  def values(schema, field) do
-    schema
+  @doc "Returns the possible values for a given schema_or_changeset and field"
+  @spec values(Ecto.Schema.t | Ecto.Changeset.t, atom) :: [atom()]
+  def values(schema_or_changeset, field) do
+    schema_or_changeset
     |> mappings(field)
     |> Keyword.keys()
   end
 
-  @doc "Returns the possible dump values for a given schema and field"
-  @spec dump_values(module, atom) :: [String.t()] | [integer()]
-  def dump_values(schema, field) do
-    schema
+  @doc "Returns the possible dump values for a given schema_or_changeset and field"
+  @spec dump_values(Ecto.Schema.t | Ecto.Changeset.t, atom) :: [String.t()] | [integer()]
+  def dump_values(schema_or_changeset, field) do
+    schema_or_changeset
     |> mappings(field)
     |> Keyword.values()
   end
 
-  @doc "Returns the mappings for a given schema and field"
-  @spec mappings(module, atom) :: Keyword.t()
-  def mappings(schema, field) do
-    try do
-      schema.__changeset__()
-    rescue
-      _ in UndefinedFunctionError ->
-        raise ArgumentError, "#{inspect(schema)} is not an Ecto schema"
-    else
+  @doc "Returns the mappings for a given schema_or_changeset and field"
+  @spec mappings(Ecto.Changeset.t, atom) :: Keyword.t()
+  def mappings(%Ecto.Changeset{types: nil}, _field) do
+    raise ArgumentError, "changeset does not have types information"
+  end
+
+  @spec mappings(Ecto.Changeset.t, atom) :: Keyword.t()
+  def mappings(%Ecto.Changeset{types: types}, field) do
+    case types do
       %{^field => {:parameterized, Ecto.Enum, %{mappings: mappings}}} -> mappings
       %{^field => {_, {:parameterized, Ecto.Enum, %{mappings: mappings}}}} -> mappings
       %{^field => _} -> raise ArgumentError, "#{field} is not an Ecto.Enum field"
       %{} -> raise ArgumentError, "#{field} does not exist"
     end
   end
+
+  @spec mappings(Ecto.Schema.t, atom) :: Keyword.t()
```

Comment:
This typespec isn't correct. `Ecto.Schema.t()` refers to a struct from a schema. This function receives a module. The same applies to the other typespecs.